### PR TITLE
Refactoring ORTTrainer tests

### PR DESCRIPTION
**Pull request by JingyaHuang**
# What does this PR do?

__Refactoring the tests of `ORTTrainer`__
- [x] Training / Inference tests (with models registered in [`transformers.onnx.features.FeaturesManager`](https://github.com/huggingface/transformers/blob/main/src/transformers/onnx/features.py#L100))
- [x] Integration tests (ex. Optuna, DeepSpeed, etc. / with a bert-like model)
- [x] Basic functionalities (ex. Hub, resume, etc.  / with a bert-like model)
- [x] Optimizer choice tests (ex. apex/ort fused adam)
- [x] Raise `NotImplementedError` in `ORTTrainer` for not implemented features.

### CODE REVIEW

File: tests/onnxruntime/nightly_test_trainer.py [hunk 1]
```diff
@@ -0,0 +1,1109 @@
...
+                max_train_samples=self.max_train_samples,
+                max_valid_samples=self.max_valid_samples,
+                max_test_samples=self.max_test_samples,
+            )
+
+            train_result = trainer.train()
+            trainer.save_model()
+            train_metrics = train_result.metrics
+            eval_metrics = trainer.evaluate(inference_with_ort=True)
+            # self.assertGreaterEqual(eval_metrics["eval_accuracy"], 0.75)
```
Comment (regisss) [inline tests/onnxruntime/nightly_test_trainer.py:443]: Should it be removed?

File: tests/onnxruntime/nightly_test_trainer.py [hunk 2]
```diff
@@ -0,0 +1,1109 @@
...
+                max_train_samples=self.max_train_samples,
+                max_valid_samples=self.max_valid_samples,
+                max_test_samples=self.max_test_samples,
+            )
+
+            train_result = trainer.train()
+            trainer.save_model()
+            train_metrics = train_result.metrics
+            eval_metrics = trainer.evaluate()
+            # self.assertGreaterEqual(eval_metrics["eval_accuracy"], 0.75)
```
Comment (regisss) [inline tests/onnxruntime/nightly_test_trainer.py:512]: Should it be removed?

File: tests/onnxruntime/nightly_test_trainer.py [hunk 3]
```diff
@@ -0,0 +1,1109 @@
...
+
+    @slow
+    @parameterized.expand(
+        _get_models_to_test(_MODELS_TO_TEST, _TASKS_DATASETS_CONFIGS, excluded=["gpt2"]), skip_on_empty=True
+    )
+    def test_trainer_fp16(self, test_name, model_name, feature, data_metric_config):
+
+        with tempfile.TemporaryDirectory() as tmp_dir:
+
+            training_args = TrainingArguments("..")
```
Comment (regisss) [inline tests/onnxruntime/nightly_test_trainer.py:495]: There should be `fp16=True` also no?

File: tests/onnxruntime/nightly_test_trainer.py [hunk 4]
```diff
@@ -0,0 +1,1109 @@
...
+                max_train_samples=self.max_train_samples,
+                max_valid_samples=self.max_valid_samples,
+                max_test_samples=self.max_test_samples,
+            )
+
+            train_result = trainer.train()
+            trainer.save_model()
+            train_metrics = train_result.metrics
+            eval_metrics = trainer.evaluate()
+            # self.assertGreaterEqual(eval_metrics["eval_accuracy"], 0.75)
```
Comment (regisss) [inline tests/onnxruntime/nightly_test_trainer.py:552]: Should it be removed?

File: tests/onnxruntime/nightly_test_trainer.py [hunk 5]
```diff
@@ -0,0 +1,1109 @@
...
```
Comment (regisss) [inline tests/onnxruntime/nightly_test_trainer.py:639]: Should it be removed?

File: tests/onnxruntime/nightly_test_trainer.py [hunk 6]
```diff
@@ -0,0 +1,1109 @@
...
```
Comment (regisss) [inline tests/onnxruntime/nightly_test_trainer.py:679]: Should it be removed?

File: tests/onnxruntime/nightly_test_trainer.py [hunk 7]
```diff
@@ -0,0 +1,1109 @@
...
+
+            train_result = trainer.train()
+            trainer.save_model()
+            train_metrics = train_result.metrics
+            ort_eval_metrics = trainer.evaluate()
+            print(ort_eval_metrics)
+            # self.assertGreaterEqual(ort_eval_metrics["eval_rouge1"], 10)
+            # self.assertGreaterEqual(ort_eval_metrics["eval_rouge2"], 2)
+            # self.assertGreaterEqual(ort_eval_metrics["eval_rougeL"], 7)
+            # self.assertGreaterEqual(ort_eval_metrics["eval_rougeLsum"], 7)
```
Comment (regisss) [inline tests/onnxruntime/nightly_test_trainer.py:1100]: Should it be removed?

**Review by regisss — APPROVED** [review]
Huge work @JingyaHuang !!
I left minor comments, mostly about commented lines that may be removed.

File: examples/onnxruntime/training/docker/run.sh [hunk 8]
```diff
@@ -16,13 +16,16 @@ python -m torch_ort.configure
...
+pip install coloredlogs absl-py rouge_score seqeval scipy sacrebleu nltk sklearn parameterized
 pip install fairscale deepspeed mpi4py
+# pip install optuna ray sigopt wandb # Hyper parameter search
 
 # Install apex
 # git clone https://github.com/NVIDIA/apex \
 #     && cd apex \
 #     && pip install -v --disable-pip-version-check --no-cache-dir --global-option="--cpp_ext" --global-option="--cuda_ext" ./
+# Inatall bitsandbytes
+# pip install bitsandbytes-cuda113
```
[the remark this answers is not in the document]
Comment (JingyaHuang) [reply examples/onnxruntime/training/docker/run.sh:28]: Yes, I prefer to keep them. They are dependencies for hyper parameter search or optimized optimizers. I put them in the comment as it is not necessary to run them every night. But in case of large release of transformers, it would be nice to uncomment them to ensure nothing is broken.

File: tests/onnxruntime/nightly_test_trainer.py [hunk 9]
```diff
@@ -0,0 +1,1109 @@
...
+                max_train_samples=self.max_train_samples,
+                max_valid_samples=self.max_valid_samples,
+                max_test_samples=self.max_test_samples,
+            )
+
+            train_result = trainer.train()
+            trainer.save_model()
+            train_metrics = train_result.metrics
+            eval_metrics = trainer.evaluate()
+            # self.assertGreaterEqual(eval_metrics["eval_accuracy"], 0.75)
```
[the remark this answers is not in the document]
Comment (JingyaHuang) [reply tests/onnxruntime/nightly_test_trainer.py:480]: The same situation, the nightly test would rather ensure that the trainer works well, so I send only a small amount of data for fine-tuning, but at some point, we might want to check the metric...

File: optimum/utils/testing_utils.py [hunk 10]
```diff
@@ -11,3 +14,41 @@ def require_hf_token(test_case):
...
         return test_case
+
+
+def require_sigopt_token_and_project(test_case):
+    """
+    Decorator marking a test that requires sigopt API token.
+    """
+    use_auth_token = os.environ.get("SIGOPT_API_TOKEN", None)
+    has_sigopt_project = os.environ.get("SIGOPT_PROJECT", None)
+    if use_auth_token or has_sigopt_project is None:
```
Comment (echarlaix) [inline optimum/utils/testing_utils.py:25]: Should it be `if use_auth_token is None or has_sigopt_project is None` ?

File: tests/onnxruntime/nightly_test_trainer.py [hunk 11]
```diff
@@ -0,0 +1,1118 @@
...
+def _get_models_to_test(model_list, task_list, excluded: Optional[List[str]] = None):
+    models_to_test = []
+
+    for name, model_name in model_list:
+        for feature, data_metric_config in task_list.items():
+            if excluded and name in excluded:
+                continue
+            models_to_test.append((f"{name}_{feature}", model_name, feature, data_metric_config))
+
+    return sorted(models_to_test)
```
Comment (echarlaix) [inline tests/onnxruntime/nightly_test_trainer.py:116]: Why do we sort `models_to_test` ?

File: tests/onnxruntime/nightly_test_trainer.py [hunk 12]
```diff
@@ -0,0 +1,1118 @@
...
+    def test_ort_fused_adam(self):
+        # Pretend that onnxruntime-training is installed and mock onnxruntime.training.optim.FusedAdam exists.
+        # Trainer.get_optimizer_cls_and_kwargs does not use FusedAdam. It only has to return the
+        # class given, so mocking onnxruntime.training.optim.FusedAdam should be fine for testing and allow
+        # the test to run without requiring an onnxruntime-training installation.
+        mock = Mock()
+        modules = {
+            "onnxruntime.training": mock,
+            "onnxruntime.training.optim": mock.optimizers,
+            " onnxruntime.training.optim.FusedAdam": mock.optimizers.FusedAdam,
```
Comment (echarlaix) [inline tests/onnxruntime/nightly_test_trainer.py:1024]:
```suggestion
            "onnxruntime.training.optim.FusedAdam": mock.optimizers.FusedAdam,
```

**Review by echarlaix — APPROVED** [review]
Huge work @JingyaHuang ! 🔥

File: optimum/utils/testing_utils.py (hunk 10 above)
Comment (JingyaHuang) [reply optimum/utils/testing_utils.py:25]: Exactly! Will fix that!

File: tests/onnxruntime/nightly_test_trainer.py (hunk 11 above)
Comment (JingyaHuang) [reply tests/onnxruntime/nightly_test_trainer.py:116]: I guess we want the error information to be ordered by alphabet to easier locate it or something 👀
